File: GamePad-Mobile/mobile/lib/core/protocol/crc16.dart
/// CRC16-CCITT with precomputed lookup table (polynomial 0x1021).
class Crc16 {
  Crc16._();

  static const int _polynomial = 0x1021;
  static const int _initialValue = 0xFFFF;

  static final List<int> _table = _buildTable();

  static List<int> _buildTable() {
    final table = List<int>.filled(256, 0);
    for (int i = 0; i < 256; i++) {
      int crc = 0;
      int val = i << 8;
      for (int j = 0; j < 8; j++) {
        if (((crc ^ val) & 0x8000) != 0) {
          crc = (crc << 1) ^ _polynomial;
        } else {
          crc <<= 1;
        }
        val <<= 1;
      }
      table[i] = crc & 0xFFFF;
    }
    return table;
  }

  /// Computes CRC16-CCITT over [data].
  static int compute(List<int> data) {
    int crc = _initialValue;
    for (final b in data) {
      final index = ((crc >> 8) ^ b) & 0xFF;
      crc = ((crc << 8) ^ _table[index]) & 0xFFFF;
    }
    return crc;
  }

  /// Appends CRC16 (big-endian) at positions [data.length - 2] and
  /// [data.length - 1]. The CRC is computed over data[0..length-3].
  static void append(List<int> data) {
    if (data.length < 2) {
      throw ArgumentError('Buffer too small for CRC16');
    }
    final crc = compute(data.sublist(0, data.length - 2));
    data[data.length - 2] = (crc >> 8) & 0xFF;
    data[data.length - 1] = crc & 0xFF;
  }

  /// Validates the CRC16 stored at the end of [data].
  static bool validate(List<int> data) {
    if (data.length < 2) return false;
    final stored = (data[data.length - 2] << 8) | data[data.length - 1];
    final computed = compute(data.sublist(0, data.length - 2));
    return stored == computed;
  }
}

File: GamePad-Mobile/pc-wpf/Protocol/Crc16.cs
namespace GamePadServer.Wpf.Protocol;

/// <summary>
/// CRC16-CCITT (polynomial 0x1021) with a precomputed lookup table.
/// </summary>
public static class Crc16
{
    private const ushort Polynomial = 0x1021;
    private const ushort InitialValue = 0xFFFF;

    private static readonly ushort[] Table;

    static Crc16()
    {
        Table = new ushort[256];
        for (ushort i = 0; i < 256; i++)
        {
            ushort crc = 0;
            ushort val = (ushort)(i << 8);
            for (int j = 0; j < 8; j++)
            {
                if (((crc ^ val) & 0x8000) != 0)
                    crc = (ushort)((crc << 1) ^ Polynomial);
                else
                    crc <<= 1;
                val <<= 1;
            }
            Table[i] = crc;
        }
    }

    /// <summary>
    /// Computes CRC16-CCITT over the given byte span.
    /// </summary>
    public static ushort Compute(ReadOnlySpan<byte> data)
    {
        ushort crc = InitialValue;
        foreach (byte b in data)
        {
            byte index = (byte)((crc >> 8) ^ b);
            crc = (ushort)((crc << 8) ^ Table[index]);
        }
        return crc;
    }

    /// <summary>
    /// Appends CRC16 (big-endian) to the end of a packet buffer
    /// at positions [data.Length - 2] and [data.Length - 1].
    /// </summary>
    public static void Append(Span<byte> data)
    {
        if (data.Length < 2)
            throw new ArgumentException("Buffer too small for CRC16", nameof(data));

        var crc = Compute(data[..^2]);
        data[^2] = (byte)(crc >> 8);
        data[^1] = (byte)(crc);
    }

    /// <summary>
    /// Validates that the CRC16 at the end of the buffer matches
    /// the computed value over the preceding bytes.
    /// </summary>
    public static bool Validate(ReadOnlySpan<byte> data)
    {
        if (data.Length < 2)
            return false;

        var stored = (ushort)((data[^2] << 8) | data[^1]);
        var computed = Compute(data[..^2]);
        return stored == computed;
    }
}
